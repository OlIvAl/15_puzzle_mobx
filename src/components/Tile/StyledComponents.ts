import styled from 'styled-components';
import {BORD_SIZE, TILE_SIZE} from '../../constants/config';

interface IStyledProps {
  top: number;
  left: number;
}

export const StyledTile = styled('div')<IStyledProps>`
  cursor: pointer;
  background-color: #fff;
  color: #5f5f5f;
  border: 1px solid #aaa;
  display: flex;
  align-items: center;
  justify-content: center;
  font-size: 36px;
  box-sizing: border-box;
  width: ${TILE_SIZE}px;
  height: ${TILE_SIZE}px;
  position: absolute;
  top: ${({top}: IStyledProps): number => top}px;
  left: ${({left}: IStyledProps): number => left}px;
  
  @media (hover: hover) {
    &:hover {
      background: #f5f5f5;
    }
  }
  
  @media (max-width: ${BORD_SIZE * 1.2}px) {
    width: ${TILE_SIZE / 1.5}px;
    height: ${TILE_SIZE / 1.5}px;
  }
`;