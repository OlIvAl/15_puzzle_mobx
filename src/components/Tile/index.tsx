import React from 'react';
import {BORD_SIZE} from '../../constants/config';
import {StyledTile} from './StyledComponents';

import {ITile} from '../../stores/GameStore/interface';

interface IProps extends Pick<ITile, 'title'> {
  top: number;
  left: number;
 //  move: IMoveTileAsyncActionCreator;
}

interface IState {
  coef: number;
}

class Tile extends React.Component<IProps, IState> {
  state: IState = {
    coef: 1
  };

  mql: MediaQueryList = window.matchMedia(`(max-width: ${BORD_SIZE * 1.2}px)`);

  constructor(props: IProps) {
    super(props);

    this.mediaQueryListener = this.mediaQueryListener.bind(this);
    //this.onClick = this.onClick.bind(this);
  }

  mediaQueryListener(event: MediaQueryListEvent): void {
    if (event.matches) {
      this.setState({coef: 1.5});
    } else {
      this.setState({coef: 1});
    }
  }

  /*onClick(): void {
    const {
      tile,
      // move
    } = this.props;

    move(tile);
  }*/

  componentDidMount(): void {
    this.mql.addListener(this.mediaQueryListener);

    if (this.mql.matches) {
      this.setState({coef: 1.5});
    }
  }

  componentWillUnmount(): void {
    this.mql.removeListener(this.mediaQueryListener);
  }

  render(): React.ReactNode {
    const {
      title,
      top,
      left
    } = this.props;

    const {coef} = this.state;

    return (
      <StyledTile
        top={top / coef}
        left={left / coef}
      >
        {title}
      </StyledTile>
    );
  }
}

export default Tile;