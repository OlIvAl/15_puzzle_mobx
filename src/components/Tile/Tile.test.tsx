import * as React from 'react';
import Tile from '.';
import {ITileWithCoords} from '../../selectors';
import {ITile} from '../../interfaces/entities';
import {IMoveTileAsyncActionCreator} from '../../interfaces/asyncActionCreators';
import {shallow} from 'enzyme';
import {shallowToJson} from 'enzyme-to-json';

interface IProps extends ITileWithCoords {
  tile: ITile;
  move: IMoveTileAsyncActionCreator;
}

interface IState {
  coef: number;
}

const TITLE: number = 1;
const TOP: number = 0;
const LEFT: number = 0;
const ROW: number = 0;
const COL: number = 0;

describe('Tile', () => {
  beforeAll(() => {
    Object.defineProperty(window, "matchMedia", {
      value: jest.fn(() => ({
        matches: false,
        addListener: () => {},
        removeListener: () => {}
      }))
    });
  });

  const resizeWindow = (x: number, y: number) => {
    // @ts-ignore
    window.innerWidth = x;
    // @ts-ignore
    window.innerHeight = y;
    window.dispatchEvent(new Event('resize'));
  };

  const tile: ITile = {
    title: TITLE,
    col: COL,
    row: ROW
  };

  const tileCoord: ITileWithCoords = {
    title: TITLE,
    top: TOP,
    left: LEFT
  };

  const move = jest.fn() as IMoveTileAsyncActionCreator;

  it('он отображается', () => {

    const wrapper = shallow<IProps, IState>(
      <Tile
        title={tileCoord.title}
        left={tileCoord.left}
        top={tileCoord.top}
        move={move}
        tile={tile}
      />
    );

    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('при клике вызовится move', () => {
    const wrapper = shallow<IProps, IState>(
      <Tile
        title={tileCoord.title}
        left={tileCoord.left}
        top={tileCoord.top}
        move={move}
        tile={tile}
      />
    );

    wrapper.simulate('click');

    expect(move).toHaveBeenCalledWith(tile);
  });

  it('коэффициент на широком экране = 1', () => {
    const wrapper = shallow<IProps, IState>(
      <Tile
        title={tileCoord.title}
        left={tileCoord.left}
        top={tileCoord.top}
        move={move}
        tile={tile}
      />
    );

    resizeWindow(1000, 1000);

    expect(wrapper.state().coef).toEqual(1);
  });

  /*it('коэффициент на узком экране = 1,5', () => {
    const wrapper = shallow<IProps, IState>(
      <Tile
        title={tileCoord.title}
        left={tileCoord.left}
        top={tileCoord.top}
        move={move}
        tile={tile}
      />
    );

    resizeWindow(300, 300);

    expect(wrapper.state().coef).toEqual(1.5);
  });*/

  it('содержит title, как контент', () => {
    const wrapper = shallow<IProps, IState>(
      <Tile
        title={tileCoord.title}
        left={tileCoord.left}
        top={tileCoord.top}
        move={move}
        tile={tile}
      />
    );

    expect(wrapper.contains(TITLE)).toEqual(true);
  });
});