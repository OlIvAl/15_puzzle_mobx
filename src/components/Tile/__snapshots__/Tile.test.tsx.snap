// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tile он отображается 1`] = `
<styled.div
  left={0}
  onClick={[Function]}
  top={0}
>
  1
</styled.div>
`;
